File: packages/core/src/utils/quality-trend.ts
/**
 * R3 质量趋势与回灌幂等（360 号，二轮 P0 第三件契约层）。
 *
 * 1. `buildQualityTrend`：review_metrics 行 → 章级趋势点列（升序 + 均值 +
 *    未过章清单），供 `/books/:id/quality-trend` 端点与 Dashboard 趋势卡；
 * 2. `reviewMetricContentHash` / `dedupeReplayRows`：回灌链路幂等——哈希
 *    输入不含时间戳（重放时间不同但内容相同必须命中同一指纹），对齐 ANWA
 *    ChapterArtifactSyncCheckpoint 语义；
 * 3. （promise-ledger 紧迫度见 promise-ledger.ts 扩展。）
 *
 * 双端：`engine-rs/src/utils/quality_trend.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/quality-trend-vectors.json`（TS 断言
 * `golden-quality-trend.test.ts`；Rust 差分 `tests/golden_quality_trend_diff.rs`）。
 */

export interface ReviewMetricRow {
  readonly chapter: number;
  /** 0–100 整体质量分（auditor 支持评分时才有；缺分章不进均值）。 */
  readonly overallScore?: number;
  readonly passed: boolean;
  readonly criticalCount: number;
  readonly warningCount: number;
  readonly infoCount: number;
  /** 记录时间（ISO）；不参与幂等哈希。 */
  readonly recordedAt: string;
}

export interface QualityTrendPoint {
  readonly chapter: number;
  readonly overallScore?: number;
  readonly passed: boolean;
  readonly issueCount: number;
}

export interface QualityTrend {
  readonly points: ReadonlyArray<QualityTrendPoint>;
  readonly scoredChapters: number;
  /** 有分章均值，一位小数；全部缺分时 undefined。 */
  readonly averageScore?: number;
  readonly failingChapters: ReadonlyArray<number>;
}

/** 章级趋势聚合：升序稳定排序（禁 localeCompare），缺分章保留点位但不进均值。 */
export function buildQualityTrend(rows: ReadonlyArray<ReviewMetricRow>): QualityTrend {
  const sorted = [...rows].sort((a, b) => a.chapter - b.chapter);
  const points = sorted.map((row) => ({
    chapter: row.chapter,
    overallScore: row.overallScore,
    passed: row.passed,
    issueCount: row.criticalCount + row.warningCount + row.infoCount,
  }));
  const scored = sorted.filter((row) => typeof row.overallScore === "number");
  const averageScore = scored.length > 0
    ? Math.round(
      (scored.reduce((sum, row) => sum + (row.overallScore as number), 0) / scored.length) * 10,
    ) / 10
    : undefined;
  return {
    points,
    scoredChapters: scored.length,
    averageScore,
    failingChapters: sorted.filter((row) => !row.passed).map((row) => row.chapter),
  };
}

const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const FNV_MASK = 0xffffffffffffffffn;

function fnv1aHex16(text: string): string {
  const bytes = new TextEncoder().encode(text);
  let hash = FNV_OFFSET;
  for (const byte of bytes) {
    hash ^= BigInt(byte);
    hash = (hash * FNV_PRIME) & FNV_MASK;
  }
  return hash.toString(16).padStart(16, "0");
}

/**
 * 回灌幂等指纹：FNV-1a 64（UTF-8，hex16，对齐 348 号 chunkFingerprint 口径）。
 * 输入 = chapter|score|passed|critical|warning|info（**不含 recordedAt**——
 * 重放时间不同但内容相同必须命中同一指纹）。
 */
export function reviewMetricContentHash(
  row: Omit<ReviewMetricRow, "recordedAt">,
): string {
  const score = typeof row.overallScore === "number" ? String(row.overallScore) : "-";
  return fnv1aHex16(
    `${row.chapter}|${score}|${row.passed ? 1 : 0}|${row.criticalCount}|${row.warningCount}|${row.infoCount}`,
  );
}

/** 幂等重放：按内容指纹去重（同内容仅首次 fresh；保序；时间戳不同的重复行同判）。 */
export function dedupeReplayRows(
  rows: ReadonlyArray<ReviewMetricRow>,
): { fresh: ReadonlyArray<ReviewMetricRow>; skipped: number } {
  const seen = new Set<string>();
  const fresh: ReviewMetricRow[] = [];
  for (const row of rows) {
    const { recordedAt: _recordedAt, ...content } = row;
    const hash = reviewMetricContentHash(content);
    if (seen.has(hash)) continue;
    seen.add(hash);
    fresh.push(row);
  }
  return { fresh, skipped: rows.length - fresh.length };
}

File: engine-rs/src/utils/quality_trend.rs
//! R3 质量趋势与回灌幂等（360 号，二轮 P0 第三件契约层）。
//!
//! TS 真源：`packages/core/src/utils/quality-trend.ts`；golden 唯一事实源：
//! `packages/core/src/__tests__/golden/quality-trend-vectors.json`（差分测试
//! `tests/golden_quality_trend_diff.rs` 读同一文件）。
//!
//! 移植纪律：FNV-1a 64 逐字节对齐 TS（UTF-8、hex16 零填充）；排序纯码元
//! 比较，禁 locale 感知排序；缺分章不进均值但保留点位。

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "export-bindings", derive(ts_rs::TS))]
#[cfg_attr(feature = "export-bindings", ts(export))]
#[serde(rename_all = "camelCase")]
pub struct ReviewMetricRow {
    pub chapter: i64,
    #[serde(default)]
    pub overall_score: Option<i64>,
    #[serde(default)]
    pub passed: bool,
    #[serde(default)]
    pub critical_count: i64,
    #[serde(default)]
    pub warning_count: i64,
    #[serde(default)]
    pub info_count: i64,
    #[serde(default)]
    pub recorded_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[cfg_attr(feature = "export-bindings", derive(ts_rs::TS))]
#[cfg_attr(feature = "export-bindings", ts(export))]
#[serde(rename_all = "camelCase")]
pub struct QualityTrendPoint {
    pub chapter: i64,
    /// 缺分序列化为 null（golden 期望侧用 null 表示缺失）。
    pub overall_score: Option<i64>,
    pub passed: bool,
    pub issue_count: i64,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[cfg_attr(feature = "export-bindings", derive(ts_rs::TS))]
#[cfg_attr(feature = "export-bindings", ts(export))]
#[serde(rename_all = "camelCase")]
pub struct QualityTrend {
    pub points: Vec<QualityTrendPoint>,
    pub scored_chapters: usize,
    /// 有分章均值一位小数（×10 取整后 ÷10）；全部缺分为 null。
    pub average_score: Option<f64>,
    pub failing_chapters: Vec<i64>,
}

/// 章级趋势聚合：升序稳定排序，缺分章保留点位但不进均值。
pub fn build_quality_trend(rows: &[ReviewMetricRow]) -> QualityTrend {
    let mut sorted: Vec<&ReviewMetricRow> = rows.iter().collect();
    sorted.sort_by_key(|row| row.chapter);
    let points: Vec<QualityTrendPoint> = sorted
        .iter()
        .map(|row| QualityTrendPoint {
            chapter: row.chapter,
            overall_score: row.overall_score,
            passed: row.passed,
            issue_count: row.critical_count + row.warning_count + row.info_count,
        })
        .collect();
    let scored: Vec<i64> = sorted
        .iter()
        .filter_map(|row| row.overall_score)
        .collect();
    let average_score = if scored.is_empty() {
        None
    } else {
        let sum: i64 = scored.iter().sum();
        Some(((sum as f64) / (scored.len() as f64) * 10.0).round() / 10.0)
    };
    QualityTrend {
        points,
        scored_chapters: scored.len(),
        average_score,
        failing_chapters: sorted
            .iter()
            .filter(|row| !row.passed)
            .map(|row| row.chapter)
            .collect(),
    }
}

const FNV_OFFSET: u64 = 0xcbf29ce484222325;
const FNV_PRIME: u64 = 0x100000001b3;

fn fnv1a_hex16(text: &str) -> String {
    let mut hash = FNV_OFFSET;
    for byte in text.as_bytes() {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(FNV_PRIME);
    }
    format!("{hash:016x}")
}

/// 回灌幂等指纹：FNV-1a 64（UTF-8，hex16，对齐 348 号 chunkFingerprint 口径）。
/// 输入 = chapter|score|passed|critical|warning|info（**不含 recordedAt**——
/// 重放时间不同但内容相同必须命中同一指纹）；缺分章 score 为 `-`。
pub fn review_metric_content_hash(row: &ReviewMetricRow) -> String {
    let score = row
        .overall_score
        .map(|value| value.to_string())
        .unwrap_or_else(|| "-".to_string());
    fnv1a_hex16(&format!(
        "{}|{}|{}|{}|{}|{}",
        row.chapter,
        score,
        i32::from(row.passed),
        row.critical_count,
        row.warning_count,
        row.info_count
    ))
}

/// 幂等重放结果。
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReplayDedup {
    pub fresh: Vec<ReviewMetricRow>,
    pub skipped: usize,
}

/// 幂等重放：按内容指纹去重（同内容仅首次 fresh；保序；时间戳不同的重复行同判）。
pub fn dedupe_replay_rows(rows: &[ReviewMetricRow]) -> ReplayDedup {
    let mut seen = std::collections::HashSet::new();
    let mut fresh = Vec::new();
    for row in rows {
        let hash = review_metric_content_hash(row);
        if seen.insert(hash) {
            fresh.push(row.clone());
        }
    }
    let skipped = rows.len() - fresh.len();
    ReplayDedup { fresh, skipped }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(chapter: i64, score: Option<i64>, passed: bool) -> ReviewMetricRow {
        ReviewMetricRow {
            chapter,
            overall_score: score,
            passed,
            critical_count: 0,
            warning_count: 0,
            info_count: 0,
            recorded_at: "2026-09-12T00:00:00.000Z".to_string(),
        }
    }

    #[test]
    fn average_is_one_decimal_and_unscored_rows_kept() {
        let rows = vec![row(3, Some(82), true), row(1, Some(70), false), row(2, None, true)];
        let trend = build_quality_trend(&rows);
        assert_eq!(trend.scored_chapters, 2);
        assert_eq!(trend.average_score, Some(76.0));
        assert_eq!(trend.failing_chapters, vec![1]);
        assert_eq!(trend.points.len(), 3);
        // 升序后 ch2（缺分）在 index 1。
        assert_eq!(trend.points[1].overall_score, None);
        assert_eq!(trend.points[1].chapter, 2);
    }

    #[test]
    fn hash_ignores_recorded_at() {
        let mut a = row(7, Some(82), true);
        let b = row(7, Some(82), true);
        a.recorded_at = "2020-01-01T00:00:00.000Z".to_string();
        assert_eq!(review_metric_content_hash(&a), review_metric_content_hash(&b));
    }
}
